### PR TITLE
[TEL-281] Telehealth waiting area/ practitioner tool (ios)

### PR DESCRIPTION
**Pull request by ivanjiang5628**
Follow up PR[ #30](https://github.com/janeapp/video-chat/pull/30)

## Description
- [Jira Ticket](https://janeapp.atlassian.net/browse/TEL-279)
- [Linear Ticket](https://linear.app/jane/issue/IMPT-58/telehealth-waiting-area-practitioner-tool-ios)
- [Jane Admin PR](https://github.com/janeapp/jane/pull/5912)
- [Notion](https://www.notion.so/janeapp/Telehealth-Practitioner-Tools-Waiting-Room-16330e2807294cefb8002b87a17ead3f)

This waiting area feature can be enabled/disabled in jane via the `waiting-area-enabled` option in jwt token.

- Adds `jane-waiting-area-native` feature components/reducer/actions. (based on jitsi version 4627's prejoin feature.)
- new fonts and CSS updates.
- Because socket.io client lib has a connection issue between our socket.io server and the react-native video chat app (https://github.com/socketio/socket.io-client/issues/1245), I use the `SocketWebView` web view component to receive the `remoteParticipantsStatus` info/event from jane and jane socket.io server via the video chat web app. 

### Risk / General PR Class
- [x] 💚 = Low Risk

🌟 = New Feature (Adds Functionality)
☕️ = JS Dependency Changes

### Demo Notes

Loom Video:
https://www.loom.com/share/6960e2dc17394d069378babca655f580
## Code Review
Resource: [Dev Team Notion Page](https://www.notion.so/janeapp/Dev-Team-f06c6eb2ccca4066bc63fc1ac1bd2549)
Resource: [Code Review Checklist](https://www.notion.so/janeapp/Code-Review-checklist-2c510c527ac7470c902a5e8f25f9db3c)

- [x] I clearly explained the WHY behind the work, in the Description above

#### Design
- [x] I added instructions for how to test, in the QA section below
- [ ] I added tests for changes, or determined that none were required
- [ ] I demoed this to the appropriate person
- [x] I considered both mobile & desktop views, or that wasn't relevant

#### Code
- [x] I committed code with informative git messages
- [x] I wrote readable code, or added comments if it was complex
- [x] I performed a self-review of my own code
- [x] I rebased my branch on the latest `master`
- [x] I confirmed that all CircleCI tests are passing
- [ ] I didn't add new npm packages, or else I made damn sure the `yarn.lock` changes are safe for existing production packages

### Steps to Reproduce
- install Jane Online appointment QA ios app( version 0.0.1(17) ) with test flight.
- Create an online appointment in jane admin app in iOS safari/chrome -> go to the ios video chat -> You will see the waiting area page.

### Fixed / Expected Behaviour
This waiting area feature only enables if the `telehealth-waiting-area` beta flag option is checked in jane. without the enabled beta flag, users will join the call directly.

On this waiting area page:

- When the patient opens the video chat app and is on the waiting area page, the practitioner-side video chat app will enable the admit button and the practitioner can see a `client is waiting` badge in the appointment panel in jane admin app. 
  - If the practitioner clicks on the admit button, the practitioner will join the call -> the patient-side video chat app will enable the `begin` button.
  - Then the patient can click the begin button to jump on the call and the video chat app will notify Jane that the patient has arrived.

- When both participants are on the waiting area page:
  - if the practitioner leaves the waiting area -> the patient-side will not allow the patient to join the call and the `begin` button will be disabled.
  - if the patient leaves the waiting area -> the practitioner-side will not allow the practitioner to join the call and the `Admit Client` button will be disabled.

- Older version IOS app compatibility:
  - If one participant is using the older version IOS app which doesn't include this waiting area feature update -> both the practitioner side and the patient side can start the video call by clicking the enabled `begin`/`admit client`  button.
  
### Jitsi backend update
N/A 

## Screenshots

Loom video:
https://www.loom.com/share/6960e2dc17394d069378babca655f580

### CODE REVIEW

**Review by gerbus — CHANGES_REQUESTED** [review]
Took a good look through the code here, generally looks great Ivan 👍 . I noticed a few things I thought should be addressed, and then also some thoughts/questions. Also, would love to see some screenshots of the new WaitingArea in the PR Writeup.

File: react/features/toolbox/components/native/Toolbox.js [hunk 1]
```diff
@@ -129,14 +128,16 @@ class Toolbox extends PureComponent<Props> {
...
+                    !_enableJaneWaitingAreaPage && <HangupButton
+                        styles = { hangupButtonStyles } />
+                }
                 <VideoMuteButton
                     styles = { buttonStyles }
                     toggledStyles = { toggledButtonStyles } />
-                <OverflowMenuButton
-                    styles = { buttonStylesBorderless }
-                    toggledStyles = { toggledButtonStyles } />
+                {!_enableJaneWaitingAreaPage && <OverflowMenuButton
```
[the remark this answers is not in the document]
Comment (ivanjiang5628) [reply react/features/toolbox/components/native/Toolbox.js:138]: Sorry, I should have put the inVision mockup in this PR earlier
https://projects.invisionapp.com/share/S9YDMKLGWFQ#/screens/428100003

on the waiting area page, users can only test their audio/video devices so I hid the Chat/Hangup/OverflowMenu buttons.

File: react/features/jane-waiting-area-native/index.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+export * from './actions';
+export * from './functions';
+
+export { default as JaneWaitingArea } from './components/JaneWaitingArea';
+
+import './middleware';
+import './reducer';
```
[the remark this answers is not in the document]
Comment (ivanjiang5628) [reply react/features/jane-waiting-area-native/index.js]: Hmm Seems like jitsi needs this file to inject the reducer to the global store....maybe we can keep it?

File: react/features/large-video/components/LargeVideo.native.js [hunk 3]
```diff
@@ -153,9 +167,15 @@ class LargeVideo extends Component<Props, State> {
...
 function _mapStateToProps(state) {
+    const { jwt } = state['features/base/jwt'];
+    const jwtPayload = jwt && jwtDecode(jwt) || null;
+    const participant = jwtPayload && jwtPayload.context && jwtPayload.context.user || null;
+    const participantType = participant && participant.participant_type || null;
+
     return {
         _participantId: state['features/large-video'].participantId,
-        _styles: ColorSchemeRegistry.get(state, 'LargeVideo')
+        _styles: ColorSchemeRegistry.get(state, 'LargeVideo'),
```
[the remark this answers is not in the document]
Comment (ivanjiang5628) [reply react/features/large-video/components/LargeVideo.native.js:177]: this `_styles` prop is from jitsi. probably we can keep it?

**Review by gerbus — APPROVED** [review]
Thanks for addressing all my feedback!

**Comment by gerbus** [comment]
Hey @ivanjiang5628, just curious why there were more changes? Did this PR get demo'd to @markhazlett before going to code review?

**Comment by ivanjiang5628** [comment]
> Hey @ivanjiang5628, just curious why there were more changes? Did this PR get demo'd to @markhazlett before going to code review?

@gerbus Sorry for the confusion I need to update this PR for changing the waiting message UI and fix the legacy ios app compatibility issue. I also want to remove the polling from it so I converted this PR to `draft` and will request a code review again later.

**Comment by clem-jane** [comment]
Thanks for not adding to this PR, @ivanjiang5628 
Let's keep a separate PR to add the newer changes instead of updating this PR.

**Comment by clem-jane** [comment]
This was demoed with a Testflight build linked to s14. We should be able to QA with the same setup as well. 
The key areas demoed were the patient's view and practitioner's view of the Waiting Area. The buttons only function when both participants have arrived.

**Comment by jngo-janeapp** [comment]
## Before Testing

- [ ] CI must be passing
- [ ] No conflicts
- [ ] No unresolved reviews/changes requested
- [ ] No additional changes/pushes from dev since the last round of QA
- [ ] Release note is sufficient for [#features](https://janeapp.slack.com/archives/C0KC1395G)
- [ ] Read through Notion / Jira story / Linear Issue / Slack Convo / Demo video

## Developer's Recommended Test Plan
- [ ] TBD

## Additional Thoughts on Testing
- [ ] TBD

## Mobile, Browsers, Jane Desktop

- [ ] Google Chrome
- [ ] Safari
- [ ] Firefox
- [ ] Microsoft Edge
- [ ] Jane Desktop
- [ ] iPad Mini
- [ ] iPhone

## Regression Testing / Exploratory

(Consider related areas or areas where data persists Ie. Chart Templates/Intake Forms, Chart Entries/Export to PDF, Reports / Export to CSV/xls, Payment / Receipts / Superbill(US))

- Can you export a PDF from this area/function?
- Can this functionality be reused somewhere else in Jane? Or does it share components with other areas in Jane. (Canadian & US Insurance)
- Does this area have an inverse setting? Enabled/Disabled, Beta Feature Flag
- Does this area have multiple setting types? Ie. Intake Forms, Permissions/Privacy, User types
- Does this correspond to an area in Online Booking? (Booking an appointment, Entering a credit card and paying for an appointment)
- Is this related to Payment Methods? Jane Payments: Stripe, Payfirma, Square Terminal, Cash, Gift Cards, Credit Memos, Account Credit, Packages
- Could this be affected by Family Relationships?
- Could data be affected by Merging Patient Profiles?
- Could data be affected by rebooking/canceling appointments or changing treatments?
- Can this be Created, Updated, Archived/Unarchived or Deleted? (CRUD Test)
- Activity Report?
- Trigger a validation Error?

## Peer Review for Moderate Risk or Higher
- [ ] Peer Review

## Pass Story on QA Board / Mark Done in Jira/ Mark Scheduled for Release in Linear
- [ ] LGTM 🚀

**Comment by jngo-janeapp** [comment]
## Before Testing Thursday March 25th, 2021 
- [x] CI must be passing
- [x] No conflicts
- [x] No unresolved reviews/changes requested
- [x] No additional changes/pushes from dev since the last round of QA
- [x] Release note is sufficient for [#features](https://janeapp.slack.com/archives/C0KC1395G)
- [x] Read through Notion / Jira story / Linear Issue / Slack Convo / Demo video

## Developer's Recommended Test Plan
- [x] install Jane Online appointment QA iOS app( version 1.3.0(1)) with test flight.
- [x] Create an online appointment in Jane admin app in iOS safari/chrome -> go to the iOS video chat -> You will see the waiting area page.
- [x] Create an online appointment in Jane admin app -> Go to the video chat -> You will see the waiting area page.
- [x] When the patient opens the video chat app and is on the waiting area page, the practitioner-side video chat app will 
Image from iOS (5)
enable the admit button and the practitioner can see a client is waiting badge in the appointment panel in Jane admin app.
- [x] If the practitioner clicks on the admit button, the practitioner will join the call -> the patient-side video chat app will enable the begin button.
- [x] Then the patient can click the begin button to jump on the call and the video chat app will notify Jane that the patient has arrived.
- [x] When both participants are on the waiting area page:
- [x] If the practitioner leaves the waiting area -> the patient-side will not allow the patient to join the call and the begin button will be disabled.
- [x] If the patient leaves the waiting area -> the practitioner-side will not allow the practitioner to join the call and the Admit Client button will be disabled.
- [x] Older version IOS app compatibility:
- [x] If one participant is using the older version IOS app which doesn't include this waiting area feature update -> both the practitioner side and the patient side can start the video call by clicking the enabled begin/admit client button.

## Additional Thoughts on Testing
- [x] Ensure all buttons work as expected 

## Mobile, Browsers, Jane Desktop
- [x] iPad Mini
- [x] iPhone

## Peer Review for Moderate Risk or Higher
- [x] Peer Review

## Pass Story on QA Board / Mark Done in Jira/ Mark Scheduled for Release in Linear
- [x] LGTM 🚀

**Comment by jngo-janeapp** [comment]
hey @ivanjiang5628 looks good on IOS !

**Pull request merged**